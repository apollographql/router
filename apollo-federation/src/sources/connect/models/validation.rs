/*

## @connect

- Disallowed on Subscription fields
- http:
  - Is present
  - Has one path (GET, POST, etc)
  - URL path template valid syntax
  - body is valid syntax
- selection
  - Valid syntax
- If source:
  - Matches a @source
  - Transport is the same as the @source
- If no source:
  - http:
    - Path is a fully qualified URL (plus a URL path template)
- If entity: true
  - Must be on Query type
  - Field arguments must match fields in type (and @key if present)
  - Required for types with @key (eventually with @finder)

## @source

- name:
  - Valid characters
  - Unique
- http:
  - Present
  - Valid base URL

## HTTPHeaderMapping

- name: is unique
- name: is a valid header name
- as: is a valid header name
- value: is a list of valid header values
- as: and value: cannot both be present

## Output selection

- Empty composite selection
- Leaf with selections

## Input selections (URL path templates and bodies)

- Missing $args
- Missing $this

*/
