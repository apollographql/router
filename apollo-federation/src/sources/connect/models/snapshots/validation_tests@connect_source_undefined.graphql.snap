---
source: apollo-federation/src/sources/connect/models/validation.rs
expression: "format!(\"{:?}\", errors)"
input_file: apollo-federation/src/sources/connect/models/test_data/validation/connect_source_undefined.graphql
---
[Message { code: NoSourcesDefined, message: "`@connect(source: \"v1\")` on `Query.resources` specifies a source, but none are defined. Try adding `@source(name: \"v1\")` to the schema.", locations: [Location { line: 8, column: 13 }..Location { line: 8, column: 25 }] }]
